src/pygeos.c

```diff
@@ -0,0 +1,235 @@
+#define PY_SSIZE_T_CLEAN
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+
+#include <Python.h>
+
+#define NO_IMPORT_ARRAY
+#define PY_ARRAY_UNIQUE_SYMBOL shapely_ARRAY_API
+
+#include <numpy/ndarraytypes.h>
+#include <numpy/npy_math.h>
+
+#include "geos.h"
+#include <stdio.h>
+
+
+char coordseq_identical(GEOSContextHandle_t ctx, const GEOSCoordSequence* coord_seq1,
+                        const GEOSCoordSequence* coord_seq2, int hasZ) {
+  unsigned int n1, n2, dims1, dims2;
+  double* buf1 = NULL;
+  double* buf2 = NULL;
+  npy_intp i;
+  char ret;
+
+  if (GEOSCoordSeq_getSize_r(ctx, coord_seq1, &n1) == 0) {
+    return 2;
+  }
+  if (GEOSCoordSeq_getSize_r(ctx, coord_seq2, &n2) == 0) {
+    return 2;
+  }
+  if (n1 != n2) {
+    return 0;
+  }
+
+  if (GEOSCoordSeq_getDimensions_r(ctx, coord_seq1, &dims1) == 0) {
+    return 2;
+  }
+  if (GEOSCoordSeq_getDimensions_r(ctx, coord_seq2, &dims2) == 0) {
+    return 2;
+  }
+  if (dims1 != dims2) {
+    return 0;
+  }
+
+  int hasM = 0;
+  if ((dims1 == 3) & (hasZ == 0)) {
+    hasM = 1;
+  }
+  if (dims1 == 4) {
+    hasM = 1;
+  }
+
+  buf1 = malloc(sizeof(double) * n1 * dims1);
+  if (GEOSCoordSeq_copyToBuffer_r(ctx, coord_seq1, buf1, hasZ, hasM) == 0) {
+    free(buf1);
+    return 2;
+  }
+  buf2 = malloc(sizeof(double) * n1 * dims1);
+  if (GEOSCoordSeq_copyToBuffer_r(ctx, coord_seq2, buf2, hasZ, hasM) == 0) {
+    free(buf1);
+    free(buf2);
+    return 2;
+  }
+
+  ret = 1;
+  for (i = 0; i < n1 * dims1; i++) {
+    const double a = buf1[i];
+    const double b = buf2[i];
+    if (a != b && !(npy_isnan(a) && npy_isnan(b))) {
```

Comment:
I think the advantage of the current `if (a != b && !(npy_isnan(a) && npy_isnan(b)))` is that for the identical case we already get false for `a != b`, and so don't have to check for NaNs. 
While if we do `if (a == b || (npy_isnan(a) && npy_isnan(b)))`, we would have to check for NaNs for _every_ coordinate pair, even if they are equal.

So I assume that the current implementation should be faster (I could time that to be sure)

Reply:
No need to time it or alter the function, was just a suggestion.

Timing wise, wouldn't `a==b` short circuit out of the OR if true?  That was my assumption, which is why I thought these would be equivalent in timing and only different in readability...

Reply:
Ah, yes, of course, that should just as well short circuit :facepalm:

Reply:
FWIW, I didn't update for this comment because "not equal until proven equal" actually doesn't work with `break`ing early, as you need to check until the last coordinate to be sure they all are equal.